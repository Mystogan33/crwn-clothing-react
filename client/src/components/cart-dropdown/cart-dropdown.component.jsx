import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';

import { createStructuredSelector } from 'reselect';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import { CartItem } from '../cart-item/cart-item.component';
import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer
} from './cart-dropdown.styles';

const CartDropdownCmp = React.memo(
  ({ cartItems, toggleCartHidden, history }) => (
      <CartDropdownContainer>
        <CartItemsContainer>
          { cartItems.length
            ? cartItems.map(item => <CartItem key={item.id} item={item} />)
            : <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
          }
        </CartItemsContainer>
        <CartDropdownButton
          onClick={() => {
            toggleCartHidden();
            history.push('/checkout');
          }}>GO TO CHECKOUT</CartDropdownButton>
      </CartDropdownContainer>
  )
);


const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});

const mapDispatchToProps = dispatch => ({
  toggleCartHidden: () => dispatch(toggleCartHidden())
});


export const CartDropdown = compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(CartDropdownCmp);
