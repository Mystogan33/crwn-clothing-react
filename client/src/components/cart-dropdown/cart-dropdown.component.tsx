import React from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { compose } from 'redux';

import { createStructuredSelector } from 'reselect';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import { CartItem } from '../cart-item/cart-item.component';
import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer
} from './cart-dropdown.styles';

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});

const connector = connect(mapStateToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;
type ICartDropdownProps = PropsFromRedux & {
  history: any
};

export const CartDropdownCmp: React.FC<ICartDropdownProps> = ({ cartItems, dispatch , history }) => (
  <CartDropdownContainer>
    <CartItemsContainer>
      { cartItems.length
        ? cartItems.map((item: any) => <CartItem key={item.id} item={item} />)
        : <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
      }
    </CartItemsContainer>
    <CartDropdownButton
      onClick={() => {
        dispatch(toggleCartHidden());
        history.push('/checkout');
      }}>GO TO CHECKOUT</CartDropdownButton>
  </CartDropdownContainer>
);

export const CartDropdown = compose(
  withRouter,
  connector
)(CartDropdownCmp) as React.ComponentType<any>;
