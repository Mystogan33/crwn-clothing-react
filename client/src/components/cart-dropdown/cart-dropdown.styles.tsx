import styled from 'styled-components';
import { CustomButton } from '../custom-button/custom-button.component';

const CartDropdownContainer = styled.div`
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;
  border-radius: 0.3rem;
`;

const CartDropdownButton = styled(CustomButton)`
  margin-top: auto;
`;

const EmptyMessageContainer = styled.span`
  font-size: 18px;
  margin: 50px auto;
`;

const CartItemsContainer = styled.div`
  height: 240px;
  display: flex;
  flex-direction: column;
  overflow: auto;
`;

export {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer
};
