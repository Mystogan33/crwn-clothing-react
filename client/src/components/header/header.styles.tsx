import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';

const hoverOrActiveLink = css`
  color: grey;
  &:before,
  &:after {
    width: 50%;
    opacity: 1;
  }
`;

// Components

const HeaderContainer = styled.div`
  height: 70px;
  width: 100%;
  display: flex;
  justify-content: space-between;
  margin-bottom: 25px;
  padding: 0 40px;
  transition: all .5s ease-in-out, box-shadow 0s;

  &.scrolled {
    padding: 0 20px;
    position: sticky;
    top: 0;
    opacity: 0.96;
    background-color: white;
    z-index: 1;
    box-shadow: 0px 7px 24px 4px rgba(0,0,0,0.2);
  }

  @media screen and (max-width: 800px) {
    padding: 0 20px;
  }
`;

const LogoContainer = styled(Link)`
  height: 100%;
  padding: 14px 5px;
`;

const OptionsContainer = styled.div`
  width: 50%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  text-shadow: 1px 1px 2px pink;

  @media screen and (max-width: 800px) {
    width: 80%;
  }
`;

const isActive = ({ to, active }: { to?: string, active?: string}) => {
  if(to && (to === active))
    return hoverOrActiveLink;
};

const OptionLink = styled(Link)`
  padding: 10px 15px;
  cursor: pointer;
  font-weight: bold;
  transition: all 0.2s ease-in-out;
  position: relative;

  &:before,
  &:after {
    content: "";
    position: absolute;
    bottom: -2px;
    width: 0px;
    height: 2px;
    margin: 2px 0 0;
    transition: all 0.5s ease-in-out;
    opacity: 0;
    background-color: black;
  }
  &:before {
    left: 50%;
  }
  &:after {
    right: 50%;
  }

  ${isActive}

  &:hover {
    ${hoverOrActiveLink}
  }
`;

export {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink
}
