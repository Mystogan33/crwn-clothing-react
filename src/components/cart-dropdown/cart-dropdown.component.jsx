import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { createStructuredSelector } from 'reselect';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import { CartItem } from '../cart-item/cart-item.component';
import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer
} from './cart-dropdown.styles';

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});

const mapDispatchToProps = dispatch => ({
  toggleCartHidden: () => dispatch(toggleCartHidden())
});

export const CartDropdown = withRouter(connect(mapStateToProps, mapDispatchToProps)(
  ({ cartItems, toggleCartHidden, history }) => (
    <CartDropdownContainer>
      <CartItemsContainer>
        {
          cartItems.length
          ? cartItems.map(item => <CartItem key={item.id} item={item} />)
          : <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
        }
      </CartItemsContainer>
      <CartDropdownButton
        onClick={() => {
          toggleCartHidden();
          history.push('/checkout');
        }}>GO TO CHECKOUT</CartDropdownButton>
    </CartDropdownContainer>
  )
));
