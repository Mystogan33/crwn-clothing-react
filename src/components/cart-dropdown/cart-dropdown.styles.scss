.cart-dropdown {
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;
  border-radius: 0.3rem;
  user-select: none;

  .cart-items {
    height: 240px;
    display: flex;
    flex-direction: column;
    overflow: scroll;
  }

  button {
    margin-top: auto;
  }
}
