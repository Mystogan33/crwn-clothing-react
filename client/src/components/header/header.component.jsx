import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { withRouter } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';

import { selectCartHidden } from '../../redux/cart/cart.selectors';
import { selectCurrentUser } from '../../redux/user/user.selectors';
import { signOutStart } from '../../redux/user/user.actions';

import { CartIcon } from '../cart-icon/cart-icon.component';
import { CartDropdown } from '../cart-dropdown/cart-dropdown.component';

import { ReactComponent as Logo } from '../../assets/crown.svg';

import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink
} from './header.styles';

export const HeaderComponent = ({
  currentUser,
  hidden,
  location: { pathname },
  signOutStart }) => {

  const [showNavbar, setShowNavbar] = useState(false);

  useEffect(() => {
    const scrollListener = (event) => {
      if(window.scrollY <= 75 && showNavbar === true) {
        setShowNavbar(false);
      } else if(window.scrollY > 75 && showNavbar === false) {
        setShowNavbar(true);
      }
    };

    window.addEventListener('scroll', scrollListener);
    return () => window.removeEventListener('scroll', scrollListener);
  });

    return (
      <HeaderContainer className={`${showNavbar === true ? 'scrolled': '' }`}>
        <LogoContainer to="/">
          <Logo className="logo" />
        </LogoContainer>
        <OptionsContainer>
          <OptionLink to="/shop" active={pathname}>SHOP</OptionLink>
          <OptionLink to="/contact" active={pathname}>CONTACT</OptionLink>
          { currentUser
            ? <OptionLink as='div' onClick={signOutStart}> SIGN OUT </OptionLink>
            : <OptionLink to="/signIn" active={pathname}> SIGN IN </OptionLink>
          }
          <CartIcon />
        </OptionsContainer>
        {
          hidden ? null : <CartDropdown />
        }
      </HeaderContainer>
    )
};

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
});

const mapDispatchToProps = dispatch => ({
  signOutStart: () => dispatch(signOutStart())
});

export const Header = compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(HeaderComponent);
