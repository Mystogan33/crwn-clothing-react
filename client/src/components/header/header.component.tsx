import React, { useEffect, useState } from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { Dispatch } from 'redux';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';

import { RootState } from '../../redux/root-reducer';

import { selectCartHidden } from '../../redux/cart/cart.selectors';
import { selectCurrentUser } from '../../redux/user/user.selectors';
import { signOutStart } from '../../redux/user/user.actions';

import { CartIcon } from '../cart-icon/cart-icon.component';
import { CartDropdown } from '../cart-dropdown/cart-dropdown.component';
import { User } from '../../interfaces/interfaces';

import { ReactComponent as Logo } from '../../assets/crown.svg';

import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink
} from './header.styles';

interface Selectors {
  currentUser: User | null,
  hidden: boolean
};

const mapStateToProps = createStructuredSelector<RootState, Selectors>({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  signOutStart: () => dispatch(signOutStart())
});

const connector = connect(mapStateToProps, mapDispatchToProps);
type ReduxProps = ConnectedProps<typeof connector>;
type HeaderProps = ReduxProps & RouteComponentProps;

export const HeaderComponent = ({
  currentUser,
  hidden,
  location: { pathname },
  signOutStart }: HeaderProps) => {

  const [showNavbar, setShowNavbar] = useState(false);

  useEffect(() => {
    const scrollListener = (event: Event) => {
      if(window.scrollY <= 75 && showNavbar === true) {
        setShowNavbar(false);
      } else if(window.scrollY > 75 && showNavbar === false) {
        setShowNavbar(true);
      }
    };

    window.addEventListener('scroll', scrollListener);
    return () => window.removeEventListener('scroll', scrollListener);
  });

    return (
      <HeaderContainer className={`${showNavbar === true ? 'scrolled': '' }`}>
        <LogoContainer to="/">
          <Logo className="logo" />
        </LogoContainer>
        <OptionsContainer>
          <OptionLink to="/shop" active={pathname}>SHOP</OptionLink>
          <OptionLink to="/contact" active={pathname}>CONTACT</OptionLink>
          { currentUser
            ? <OptionLink as='div' onClick={signOutStart}>SIGN OUT</OptionLink>
            : <OptionLink to="/signIn" active={pathname}>SIGN IN</OptionLink>
          }
          <CartIcon />
        </OptionsContainer>
        {
          hidden ? null : <CartDropdown />
        }
      </HeaderContainer>
    )
};

const connectedComponent = connector(HeaderComponent);
export const Header = withRouter(connectedComponent);
