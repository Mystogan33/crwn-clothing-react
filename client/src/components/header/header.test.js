import React from 'react';
import { shallow } from 'enzyme';

import { HeaderComponent } from './header.component';
import { OptionLink } from './header.styles';
import { CartDropdown } from '../cart-dropdown/cart-dropdown.component';

describe('Header component', () => {
  let wrapper;
  let mockSignOutStart;
  let mockLocation = { pathname: '/' };

  beforeEach(() => {
    mockSignOutStart = jest.fn();
    const mockCurrentUser = { uid: '123' };
    const mockProps = {
      hidden: true,
      currentUser: mockCurrentUser,
      signOutStart: mockSignOutStart,
      location: mockLocation
    };

    wrapper = shallow(<HeaderComponent {...mockProps} />);
  });

  it('should render Header Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('if currentUser is present', () => {
    it('should render sign out link', () => {
      expect(wrapper.find(OptionLink).at(2).text()).toBe('SIGN OUT');
    });

    it('should call signOutStart method when link is clicked', () => {
      wrapper.find(OptionLink).at(2).simulate('click');
      expect(mockSignOutStart).toHaveBeenCalled();
    });
  });

  describe('if currentUser is null', () => {
    it('should render sign in link', () => {

      const mockProps = {
        hidden: true,
        currentUser: null,
        signOutStart: mockSignOutStart,
        location: mockLocation
      };

      const newWrapper = shallow(<HeaderComponent {...mockProps} />);
      expect(newWrapper.find(OptionLink).at(2).text()).toBe('SIGN IN');
    });
  });

  describe('if hidden is true', () => {
    it('should not render CartDropdown', () => {

      const mockProps = {
        hidden: true,
        currentUser: null,
        signOutStart: mockSignOutStart,
        location: mockLocation
      };

      const newWrapper = shallow(<HeaderComponent {...mockProps} />);
      expect(newWrapper.exists(CartDropdown)).toBe(false);
    });
  });

  describe('if hidden is false', () => {
    it('should render CartDropdown', () => {

      const mockProps = {
        hidden: false,
        currentUser: null,
        signOutStart: mockSignOutStart,
        location: mockLocation
      };

      const newWrapper = shallow(<HeaderComponent {...mockProps } />);
      expect(newWrapper.exists(CartDropdown)).toBe(true);
    });
  });
});