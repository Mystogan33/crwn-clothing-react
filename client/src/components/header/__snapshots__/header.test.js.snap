// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render Header Component 1`] = `ShallowWrapper {}`;
